base/expr.jl

```diff
@@ -753,6 +752,39 @@ function compute_assumed_setting(@nospecialize(setting), val::Bool=true)
     end
 end
 
+"""
+    Base.@noinfer function f(args...)
+        @nospecialize ...
+        ...
+    end
+    Base.@noinfer f(@nospecialize args...) = ...
+
+Tells the compiler to infer `f` using the declared types of `@nospecialize`d arguments.
+This can be used to limit the number of compiler-generated specializations during inference.
+
+# Example
+
+```julia
+julia> f(A::AbstractArray) = g(A)
+f (generic function with 1 method)
+
+julia> @noinline Base.@noinfer g(@nospecialize(A::AbstractArray)) = A[1]
+g (generic function with 1 method)
+
+julia> @code_typed f([1.0])
+CodeInfo(
+1 ─ %1 = invoke Main.g(_2::AbstractArray)::Any
+└──      return %1
+) => Any
+```
+
+In this example, `f` will be inferred for each specific type of `A`,
+but `g` will only be inferred once.
```

Comment:
Would it make sense to add something like this? "Without the `@noinfer`, `f([1.0])` would infer the return type of `g` as `Float64`, a hint that inference ran for `g(::Vector{Float64})` despite the prohibition on specialized code generation."

Reply:
Added!